Liberty Fails to Load Tag Files from META-INF/resources/ from JARs

**Describe the bug**  
If a jar packaged under WEB-INF/lib contains tag files (located within `META-INF/resources/WEB-INF/tags`) then Liberty fails to load the tag files. 

Error displayed in the logs: `JSPG0046E: Unable to locate tagfile: myTag` 

> com.ibm.ws.jsp.webcontainerext.JSPErrorReport: /tag-test.jsp&#40;11,1&#41; --&gt; JSPG0046E: Unable to locate tagfile for tag myTag 
> 	at com.ibm.ws.jsp.translator.visitor.validator.ValidateVisitor.getTagInfo(ValidateVisitor.java:1928)
> 	at [internal classes]
> Caused by: com.ibm.ws.jsp.translator.JspTranslationException: /tag-test.jsp(11,1) --> JSPG0046E: Unable to locate tagfile for tag myTag 
> 	... 3 more


**Steps to Reproduce**  

1) Create a JAR file containing tag files at META-INF/resources/WEB-INF/tags/

my-taglib.jar
└── META-INF/
    └── resources/
        └── WEB-INF/
            └── tags/
                └── myTag.tag

2) Place the JAR in your WAR's /WEB-INF/lib/ directory

3) Create a JSP that uses the tag file:

<%@ taglib tagdir="/WEB-INF/tags" prefix="custom" %>
<custom:myTag/>

**Expected behavior**  
The JSP should compile and execute successfully, loading the tag file from the JAR.

**Diagnostic information:**  
 - OpenLiberty Version: ALL
 - Affected feature(s) jsp-2.2, jsp-2.3, pages-3.0, pages-3.1
 - Java Version: N/A

**Additional context**  
Workaround: Unzip the tags from the jar, and place them within the WAR WEB-INF/tags directory instead. 

Tested this scenario on apache-tomcat-10.1.45, and, by default, the tags files are picked up automatically. 

See Skill Case TS021883410

--- a/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/FATSuite.java
+++ b/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/FATSuite.java
@@ -32,6 +32,7 @@ import com.ibm.ws.jsp23.fat.tests.JSPJava7Test;
 import com.ibm.ws.jsp23.fat.tests.JSPJava8Test;
 import com.ibm.ws.jsp23.fat.tests.JSPPrepareJSPThreadCountDefaultValueTests;
 import com.ibm.ws.jsp23.fat.tests.JSPPrepareJSPThreadCountNonDefaultValueTests;
+import com.ibm.ws.jsp23.fat.tests.JSPTagFilesInJarTest;
 import com.ibm.ws.jsp23.fat.tests.JSTLTests;
 
 import componenttest.rules.repeater.FeatureReplacementAction;
@@ -60,7 +61,8 @@ import componenttest.topology.impl.LibertyServerFactory;
                 JSPChannelTest.class,
                 JSPDebugSupport.class,
                 JSPExpressionLanguageTests.class,
-                JSPEmptyIncludePathTest.class
+                JSPEmptyIncludePathTest.class,
+                JSPTagFilesInJarTest.class
 })
 
 public class FATSuite {

--- a/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPEmptyIncludePathTest.java
+++ b/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPEmptyIncludePathTest.java
@@ -29,6 +29,8 @@ import componenttest.annotation.ExpectedFFDC;
 import componenttest.annotation.Server;
 import componenttest.annotation.SkipForRepeat;
 import componenttest.custom.junit.runner.FATRunner;
+import componenttest.custom.junit.runner.Mode;
+import componenttest.custom.junit.runner.Mode.TestMode;
 import componenttest.topology.impl.LibertyServer;
 
 /**
@@ -46,6 +48,7 @@ import componenttest.topology.impl.LibertyServer;
  */
 @SkipForRepeat("CDI-2.0")
 @RunWith(FATRunner.class)
+@Mode(TestMode.FULL)
 public class JSPEmptyIncludePathTest {
     private static final String APP_NAME = "TestEmptyIncludePath";
     private static final Logger LOG = Logger.getLogger(JSPEmptyIncludePathTest.class.getName());

--- a/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTagFilesInJarTest.java
+++ b/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTagFilesInJarTest.java
@@ -1,0 +1,137 @@
+/*******************************************************************************
+ * Copyright (c) 2026 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package com.ibm.ws.jsp23.fat.tests;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.io.File;
+import java.util.logging.Logger;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import com.ibm.websphere.simplicity.ShrinkHelper;
+import com.ibm.ws.jsp23.fat.JSPUtils;
+import com.meterware.httpunit.GetMethodWebRequest;
+import com.meterware.httpunit.WebConversation;
+import com.meterware.httpunit.WebRequest;
+import com.meterware.httpunit.WebResponse;
+
+import componenttest.annotation.Server;
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.topology.impl.LibertyServer;
+
+/**
+ * Test to verify that tag files packaged in a JAR under WEB-INF/lib
+ * can be loaded from META-INF/resources/WEB-INF/tags/
+ * 
+ * This test addresses the bug where Liberty fails to load tag files from JARs,
+ * throwing error: JSPG0046E: Unable to locate tagfile
+ * 
+ * The test verifies that tag files in a JAR are properly discovered and used
+ * by JSP pages, matching the behavior of Apache Tomcat.
+ */
+@RunWith(FATRunner.class)
+public class JSPTagFilesInJarTest {
+    private static final Logger LOG = Logger.getLogger(JSPTagFilesInJarTest.class.getName());
+    private static final String APP_NAME = "TestTagFilesInJar";
+
+    @Server("tagFilesInJarServer")
+    public static LibertyServer server;
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        // Create a JAR with tag files in META-INF/resources/WEB-INF/tags/
+        JavaArchive tagLibJar = ShrinkWrap.create(JavaArchive.class, "test-taglib.jar")
+                        .addAsResource(new File("test-applications/TestTagFilesInJar.war/test-taglib/META-INF/resources/WEB-INF/tags/myTag.tag"),
+                                       "META-INF/resources/WEB-INF/tags/myTag.tag")
+                        .addAsResource(new File("test-applications/TestTagFilesInJar.war/test-taglib/META-INF/resources/WEB-INF/tags/anotherTag.tag"),
+                                       "META-INF/resources/WEB-INF/tags/anotherTag.tag");
+
+        // Create the WAR and add the JAR to WEB-INF/lib
+        WebArchive app = ShrinkWrap.create(WebArchive.class, APP_NAME + ".war")
+                        .addAsWebInfResource(new File("test-applications/TestTagFilesInJar.war/resources/WEB-INF/web.xml"))
+                        .addAsWebResource(new File("test-applications/TestTagFilesInJar.war/resources/tag-test.jsp"))
+                        .addAsWebResource(new File("test-applications/TestTagFilesInJar.war/resources/multiple-tags-test.jsp"))
+                        .addAsLibrary(tagLibJar);
+
+        ShrinkHelper.exportDropinAppToServer(server, app);
+
+        server.startServer(JSPTagFilesInJarTest.class.getSimpleName() + ".log");
+    }
+
+    @AfterClass
+    public static void testCleanup() throws Exception {
+        if (server != null && server.isStarted()) {
+            server.stopServer();
+        }
+    }
+
+    /**
+     * Test that a JSP can successfully use a tag file packaged in a JAR
+     * located in WEB-INF/lib with the tag file at META-INF/resources/WEB-INF/tags/
+     * 
+     * @throws Exception if the test fails
+     */
+    @Test
+    public void testTagFileFromJar() throws Exception {
+        WebConversation wc = new WebConversation();
+        wc.setExceptionsThrownOnErrorStatus(false);
+
+        String url = JSPUtils.createHttpUrlString(server, APP_NAME, "tag-test.jsp");
+        LOG.info("url: " + url);
+
+        WebRequest request = new GetMethodWebRequest(url);
+        WebResponse response = wc.getResponse(request);
+        LOG.info("JSP response: " + response.getText());
+
+        assertEquals("Expected 200 status code was not returned!",
+                     200, response.getResponseCode());
+        
+        // Verify the tag was successfully invoked and rendered
+        assertTrue("The response did not contain expected output from tag file",
+                   response.getText().contains("Hello from myTag!"));
+        assertTrue("The response did not contain the message parameter",
+                   response.getText().contains("Test message from JSP"));
+    }
+
+    /**
+     * Test that a JSP can use multiple tag files from the same JAR
+     * 
+     * @throws Exception if the test fails
+     */
+    @Test
+    public void testMultipleTagFilesFromJar() throws Exception {
+        WebConversation wc = new WebConversation();
+        wc.setExceptionsThrownOnErrorStatus(false);
+
+        String url = JSPUtils.createHttpUrlString(server, APP_NAME, "multiple-tags-test.jsp");
+        LOG.info("url: " + url);
+
+        WebRequest request = new GetMethodWebRequest(url);
+        WebResponse response = wc.getResponse(request);
+        LOG.info("JSP response: " + response.getText());
+
+        assertEquals("Expected 200 status code was not returned!",
+                     200, response.getResponseCode());
+        
+        // Verify both tags were successfully invoked
+        assertTrue("The response did not contain output from myTag",
+                   response.getText().contains("Hello from myTag!"));
+        assertTrue("The response did not contain output from anotherTag",
+                   response.getText().contains("Another tag output"));
+    }
+}

--- a/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTagFilesInJarTest.java
+++ b/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTagFilesInJarTest.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2026 IBM Corporation and others.
+ * Copyright 2026 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
@@ -24,6 +24,7 @@ import org.junit.Test;
 import org.junit.runner.RunWith;
 
 import com.ibm.websphere.simplicity.ShrinkHelper;
+import com.ibm.websphere.simplicity.config.ServerConfiguration;
 import com.ibm.ws.jsp23.fat.JSPUtils;
 import com.meterware.httpunit.GetMethodWebRequest;
 import com.meterware.httpunit.WebConversation;
@@ -76,18 +77,33 @@ public class JSPTagFilesInJarTest {
     @AfterClass
     public static void testCleanup() throws Exception {
         if (server != null && server.isStarted()) {
-            server.stopServer();
+             server.stopServer("SRVE8115W", "SRVE8094W");
         }
     }
 
     /**
      * Test that a JSP can successfully use a tag file packaged in a JAR
      * located in WEB-INF/lib with the tag file at META-INF/resources/WEB-INF/tags/
-     * 
+     *
      * @throws Exception if the test fails
      */
     @Test
     public void testTagFileFromJar() throws Exception {
+        // Enable loadTagFilesFromJars in server configuration
+        ServerConfiguration configuration = server.getServerConfiguration();
+        configuration.getJspEngine().setLoadTagFilesFromJars(true);
+        LOG.info("Enabling loadTagFilesFromJars: " + configuration);
+
+        server.setMarkToEndOfLog();
+        server.updateServerConfiguration(configuration);
+
+        // Wait for the server configuration update to complete before restarting the application.
+        server.waitForConfigUpdateInLogUsingMark(null);
+
+        // Restart the application and ensure it finishes starting.
+        server.restartApplication(APP_NAME);
+        server.waitForStringInLogUsingMark("CWWKT0016I:.*" + APP_NAME + ".*");
+
         WebConversation wc = new WebConversation();
         wc.setExceptionsThrownOnErrorStatus(false);
 
@@ -110,11 +126,26 @@ public class JSPTagFilesInJarTest {
 
     /**
      * Test that a JSP can use multiple tag files from the same JAR
-     * 
+     *
      * @throws Exception if the test fails
      */
     @Test
     public void testMultipleTagFilesFromJar() throws Exception {
+        // Enable loadTagFilesFromJars in server configuration
+        ServerConfiguration configuration = server.getServerConfiguration();
+        configuration.getJspEngine().setLoadTagFilesFromJars(true);
+        LOG.info("Enabling loadTagFilesFromJars: " + configuration);
+
+        server.setMarkToEndOfLog();
+        server.updateServerConfiguration(configuration);
+
+        // Wait for the server configuration update to complete before restarting the application.
+        server.waitForConfigUpdateInLogUsingMark(null);
+
+        // Restart the application and ensure it finishes starting.
+        server.restartApplication(APP_NAME);
+        server.waitForStringInLogUsingMark("CWWKT0016I:.*" + APP_NAME + ".*");
+
         WebConversation wc = new WebConversation();
         wc.setExceptionsThrownOnErrorStatus(false);
 
@@ -133,5 +164,57 @@ public class JSPTagFilesInJarTest {
                    response.getText().contains("Hello from myTag!"));
         assertTrue("The response did not contain output from anotherTag",
                    response.getText().contains("Another tag output"));
+    }
+
+    /**
+     * Test that tag files in JARs are NOT loaded when loadTagFilesFromJars is set to false.
+     * This should result in an error when trying to use the tag.
+     *
+     * @throws Exception if the test fails
+     */
+    @Test
+    public void testTagFileFromJarWithLoadTagFilesFromJarsDisabled() throws Exception {
+        // Disable loadTagFilesFromJars in server configuration
+        ServerConfiguration configuration = server.getServerConfiguration();
+        configuration.getJspEngine().setLoadTagFilesFromJars(false);
+        LOG.info("Disabling loadTagFilesFromJars: " + configuration);
+
+        server.setMarkToEndOfLog();
+        server.updateServerConfiguration(configuration);
+
+        // Wait for the server configuration update to complete before restarting the application.
+        server.waitForConfigUpdateInLogUsingMark(null);
+
+        // Restart the server to pick up the new config.
+        server.stopServer();
+        server.startServer(JSPTagFilesInJarTest.class.getSimpleName() + "-2.log");
+        server.waitForStringInLogUsingMark("CWWKT0016I:.*" + APP_NAME + ".*");
+
+        WebConversation wc = new WebConversation();
+        wc.setExceptionsThrownOnErrorStatus(false);
+
+        String url = JSPUtils.createHttpUrlString(server, APP_NAME, "tag-test.jsp");
+        LOG.info("url: " + url);
+
+        WebRequest request = new GetMethodWebRequest(url);
+        WebResponse response = wc.getResponse(request);
+        LOG.info("JSP response: " + response.getText());
+
+        // When loadTagFilesFromJars is false, the tag file should not be found
+        // and we should get a 500 error or the page should contain an error message
+        assertTrue("Expected error when loadTagFilesFromJars is false",
+                   response.getResponseCode() == 500 ||
+                   response.getText().contains("JSPG0046E") ||
+                   response.getText().contains("Unable to locate tagfile"));
+
+        // Re-enable loadTagFilesFromJars for subsequent tests
+        configuration.getJspEngine().setLoadTagFilesFromJars(true);
+        LOG.info("Re-enabling loadTagFilesFromJars: " + configuration);
+
+        server.setMarkToEndOfLog();
+        server.updateServerConfiguration(configuration);
+        server.waitForConfigUpdateInLogUsingMark(null);
+        server.restartApplication(APP_NAME);
+        server.waitForStringInLogUsingMark("CWWKT0016I:.*" + APP_NAME + ".*");
     }
 }

--- a/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTagFilesInJarTest.java
+++ b/dev/com.ibm.ws.jsp.2.3_fat/fat/src/com/ibm/ws/jsp23/fat/tests/JSPTagFilesInJarTest.java
@@ -33,6 +33,8 @@ import com.meterware.httpunit.WebResponse;
 
 import componenttest.annotation.Server;
 import componenttest.custom.junit.runner.FATRunner;
+import componenttest.custom.junit.runner.Mode;
+import componenttest.custom.junit.runner.Mode.TestMode;
 import componenttest.topology.impl.LibertyServer;
 
 /**
@@ -46,6 +48,7 @@ import componenttest.topology.impl.LibertyServer;
  * by JSP pages, matching the behavior of Apache Tomcat.
  */
 @RunWith(FATRunner.class)
+@Mode(TestMode.FULL)
 public class JSPTagFilesInJarTest {
     private static final Logger LOG = Logger.getLogger(JSPTagFilesInJarTest.class.getName());
     private static final String APP_NAME = "TestTagFilesInJar";

--- a/dev/com.ibm.ws.jsp.2.3_fat/publish/servers/tagFilesInJarServer/.gitignore
+++ b/dev/com.ibm.ws.jsp.2.3_fat/publish/servers/tagFilesInJarServer/.gitignore
@@ -1,0 +1,4 @@
+/logs/
+/workarea/
+/dropins/
+/apps/

--- a/dev/com.ibm.ws.jsp.2.3_fat/publish/servers/tagFilesInJarServer/bootstrap.properties
+++ b/dev/com.ibm.ws.jsp.2.3_fat/publish/servers/tagFilesInJarServer/bootstrap.properties
@@ -1,0 +1,13 @@
+###############################################################################
+# Copyright (c) 2026 IBM Corporation and others.
+# All rights reserved. This program and the accompanying materials
+# are made available under the terms of the Eclipse Public License 2.0
+# which accompanies this distribution, and is available at
+# http://www.eclipse.org/legal/epl-2.0/
+# 
+# SPDX-License-Identifier: EPL-2.0
+###############################################################################
+bootstrap.include=../testports.properties
+osgi.console=7777
+com.ibm.ws.logging.trace.specification=*=info:com.ibm.ws.webcontainer*=all:com.ibm.wsspi.webcontainer*=all:HTTPChannel=all:TCPChannel=all:GenericBNF=all:com.ibm.ws.jsp*=all
+

--- a/dev/com.ibm.ws.jsp.2.3_fat/publish/servers/tagFilesInJarServer/server.xml
+++ b/dev/com.ibm.ws.jsp.2.3_fat/publish/servers/tagFilesInJarServer/server.xml
@@ -1,0 +1,19 @@
+<!--
+    Copyright (c) 2026 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<server description="Server for testing JSP tag files in JAR">
+
+    <include location="../fatTestPorts.xml"/>
+
+    <featureManager>
+        <feature>jsp-2.3</feature>
+        <feature>localConnector-1.0</feature>
+    </featureManager>
+    
+</server>

--- a/dev/com.ibm.ws.jsp.2.3_fat/test-applications/TestTagFilesInJar.war/resources/WEB-INF/web.xml
+++ b/dev/com.ibm.ws.jsp.2.3_fat/test-applications/TestTagFilesInJar.war/resources/WEB-INF/web.xml
@@ -1,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2025 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<web-app
+    version="3.1"
+    xmlns="http://xmlns.jcp.org/xml/ns/javaee"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-app_3_1.xsd">
+
+  <display-name>TestTagFilesInJar</display-name>
+    
+</web-app>

--- a/dev/com.ibm.ws.jsp.2.3_fat/test-applications/TestTagFilesInJar.war/resources/multiple-tags-test.jsp
+++ b/dev/com.ibm.ws.jsp.2.3_fat/test-applications/TestTagFilesInJar.war/resources/multiple-tags-test.jsp
@@ -1,0 +1,26 @@
+<!--
+    Copyright (c) 2026 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<%@ taglib tagdir="/WEB-INF/tags" prefix="custom" %>
+<!DOCTYPE HTML>
+<html>
+<head>
+    <title>Multiple Tag Files Test</title>
+    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
+</head>
+<body>
+    <h1>Testing Multiple Tag Files from JAR</h1>
+    <div>
+        <custom:myTag message="First tag" />
+    </div>
+    <div>
+        <custom:anotherTag />
+    </div>
+</body>
+</html>

--- a/dev/com.ibm.ws.jsp.2.3_fat/test-applications/TestTagFilesInJar.war/resources/tag-test.jsp
+++ b/dev/com.ibm.ws.jsp.2.3_fat/test-applications/TestTagFilesInJar.war/resources/tag-test.jsp
@@ -1,0 +1,23 @@
+<!--
+    Copyright (c) 2026 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<%@ taglib tagdir="/WEB-INF/tags" prefix="custom" %>
+<!DOCTYPE HTML>
+<html>
+<head>
+    <title>Tag Files in JAR Test</title>
+    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
+</head>
+<body>
+    <h1>Testing Tag Files from JAR</h1>
+    <div>
+        <custom:myTag message="Test message from JSP" />
+    </div>
+</body>
+</html>

--- a/dev/com.ibm.ws.jsp.2.3_fat/test-applications/TestTagFilesInJar.war/test-taglib/META-INF/resources/WEB-INF/tags/anotherTag.tag
+++ b/dev/com.ibm.ws.jsp.2.3_fat/test-applications/TestTagFilesInJar.war/test-taglib/META-INF/resources/WEB-INF/tags/anotherTag.tag
@@ -1,0 +1,14 @@
+<!--
+    Copyright (c) 2026 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<%@ tag body-content="empty" %>
+<div class="another-tag">
+    <p>Another tag output</p>
+    <p>This is a second tag file from the JAR</p>
+</div>

--- a/dev/com.ibm.ws.jsp.2.3_fat/test-applications/TestTagFilesInJar.war/test-taglib/META-INF/resources/WEB-INF/tags/myTag.tag
+++ b/dev/com.ibm.ws.jsp.2.3_fat/test-applications/TestTagFilesInJar.war/test-taglib/META-INF/resources/WEB-INF/tags/myTag.tag
@@ -1,0 +1,15 @@
+<!--
+    Copyright (c) 2026 IBM Corporation and others.
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License 2.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-2.0/
+    
+    SPDX-License-Identifier: EPL-2.0
+ -->
+<%@ tag body-content="empty" %>
+<%@ attribute name="message" required="true" %>
+<div class="custom-tag">
+    <p>Hello from myTag!</p>
+    <p>Message: ${message}</p>
+</div>

--- a/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/JspOptions.java
+++ b/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/JspOptions.java
@@ -109,6 +109,7 @@ public class JspOptions {
     protected boolean    deleteClassFilesBeforeRecompile = false; //PI12939
     protected boolean    allowMultipleAttributeValues = false; //PI30519
     protected boolean    allowPrecedenceInJspExpressionsWithConstantString = false; //PI37304
+    protected boolean    loadTagFilesFromJars = true; // Load tag files from META-INF/resources/WEB-INF/tags in JARs
 
     //@BLB Pretouch End
     // defect 400645
@@ -662,7 +663,21 @@ public class JspOptions {
         }
         //PK57873
         
-        String ieClassId = jspParams.getProperty("ieClassId");        
+        // Load tag files from JARs
+        String loadTagFilesFromJars = jspParams.getProperty("loadTagFilesFromJars");
+        if (loadTagFilesFromJars != null) {
+            if (loadTagFilesFromJars.equalsIgnoreCase("true"))
+                this.loadTagFilesFromJars = true;
+            else if (loadTagFilesFromJars.equalsIgnoreCase("false"))
+                this.loadTagFilesFromJars = false;
+            else {
+                if(logger.isLoggable(Level.INFO)){
+                    logger.logp(Level.INFO, CLASS_NAME, "populateOptions", "Invalid value for loadTagFilesFromJars = "+ loadTagFilesFromJars);
+                }
+            }
+        }
+        
+        String ieClassId = jspParams.getProperty("ieClassId");
         if (ieClassId != null)
             this.ieClassId = ieClassId;
 
@@ -1598,6 +1613,15 @@ public class JspOptions {
         this.allowPrecedenceInJspExpressionsWithConstantString = temp;
     }
     //PI37304 end
+    
+    // Load tag files from JARs
+    public boolean isLoadTagFilesFromJars(){
+        return loadTagFilesFromJars;
+    }
+    
+    public void setLoadTagFilesFromJars(boolean temp){
+        this.loadTagFilesFromJars = temp;
+    }
 
     public String toString() {	//overrride Object's toString to assist in debugging.
     	String separatorString = System.getProperty("line.separator");

--- a/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/JspOptions.java
+++ b/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/JspOptions.java
@@ -109,7 +109,7 @@ public class JspOptions {
     protected boolean    deleteClassFilesBeforeRecompile = false; //PI12939
     protected boolean    allowMultipleAttributeValues = false; //PI30519
     protected boolean    allowPrecedenceInJspExpressionsWithConstantString = false; //PI37304
-    protected boolean    loadTagFilesFromJars = true; // Load tag files from META-INF/resources/WEB-INF/tags in JARs
+    protected boolean    loadTagFilesFromJars = false; // Load tag files from META-INF/resources/WEB-INF/tags in JARs
 
     //@BLB Pretouch End
     // defect 400645

--- a/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/TagLibraryCache.java
+++ b/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/taglib/TagLibraryCache.java
@@ -148,7 +148,7 @@ public class TagLibraryCache extends Hashtable<String, Object> {
         this.jspOptions = jspOptions; // 396002
         //outputDir = jspOptions.getOutputDir().getPath();
         if (cont!=null) {
-            container = cont;    
+            container = cont;
         } else {
             container = ctxt.getServletContext().getModuleContainer(); //null if no loose container
         }
@@ -186,6 +186,10 @@ public class TagLibraryCache extends Hashtable<String, Object> {
             if (ctxt!=null) {
                 //don't need to load these if we've been passed a container since we're just gathering classes for injection
                 loadWebInfTagFiles("/WEB-INF/tags");
+                // Also load tag files from META-INF/resources/WEB-INF/tags in JARs (if enabled)
+                if (jspOptions.isLoadTagFilesFromJars()) {
+                    loadTagFilesFromJars();
+                }
             }
         } //PK69220
 
@@ -198,7 +202,7 @@ public class TagLibraryCache extends Hashtable<String, Object> {
         }
         //PK68590 end
 
-        if (jspOptions.isUseImplicitTagLibs() && jspOptions.getTranslationContextClass() == null) { 
+        if (jspOptions.isUseImplicitTagLibs() && jspOptions.getTranslationContextClass() == null) {
             for (Map.Entry<String, Object> mapEntry : globalMap.entrySet()) {
                 String uri = mapEntry.getKey();
                 if (!containsKey(uri)) {
@@ -250,7 +254,10 @@ public class TagLibraryCache extends Hashtable<String, Object> {
     
     //only called when you have a context
     public void loadWebInfTagFiles() throws JspCoreException {
+
         loadWebInfTagFiles("/WEB-INF/tags");
+        // Also load tag files from META-INF/resources/WEB-INF/tags in JARs
+        loadTagFilesFromJars();
     }
 
     private List loadWebXmlMap(Map webxmlTagLibMap) throws JspCoreException {
@@ -706,6 +713,206 @@ public class TagLibraryCache extends Hashtable<String, Object> {
                 else if (tli!=null) {
                     put(tagsDir, tli); // just put the info from implicit.tld even though there are no tag files
                 }
+            }
+        }
+    }
+    
+    /**
+     * Loads tag files from META-INF/resources/WEB-INF/tags in JAR files located in WEB-INF/lib.
+     * This supports Servlet 3.0+ web fragments where tag files can be packaged in library JARs.
+     */
+    private void loadTagFilesFromJars() throws JspCoreException {
+        if (container != null) {
+            try {
+                NonPersistentCache npc = container.adapt(NonPersistentCache.class);
+                TagLibContainerInfo tagLibContainers = (TagLibContainerInfo) npc.getFromCache(TagLibContainerInfo.class);
+                if (tagLibContainers != null) {
+                    for (ContainerInfo info : tagLibContainers.getTagLibContainers()) {
+                        if (info.getType() == ContainerInfo.Type.WEB_INF_LIB) {
+                            loadTagFilesFromJar(info.getContainer(), info.getName());
+                        }
+                    }
+                } else {
+                    com.ibm.wsspi.adaptable.module.Entry libDir = container.getEntry("/WEB-INF/lib");
+                    if (libDir != null) {
+                        Container libDirContainer = libDir.adapt(Container.class);
+                        if (libDirContainer != null) {
+                            for (com.ibm.wsspi.adaptable.module.Entry entry : libDirContainer) {
+                                String entryName = entry.getName();
+                                if (entryName.endsWith(".jar")) {
+                                    Container jarContainer = entry.adapt(Container.class);
+                                    if (jarContainer != null) {
+                                        loadTagFilesFromJar(jarContainer, entryName);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            } catch (UnableToAdaptException e) {
+                if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.WARNING)) {
+                    logger.logp(Level.WARNING, CLASS_NAME, "loadTagFilesFromJars", "Failed to load tag files from JARs", e);
+                }
+            }
+        }
+    }
+    
+    /**
+     * Loads tag files from META-INF/resources/WEB-INF/tags in a specific JAR file.
+     * Recursively scans subdirectories under META-INF/resources/WEB-INF/tags.
+     */
+    private void loadTagFilesFromJar(Container jarContainer, String jarName) throws JspCoreException {
+        try {
+            // Look for META-INF/resources/WEB-INF/tags in the JAR
+            com.ibm.wsspi.adaptable.module.Entry metaInfResourcesEntry = jarContainer.getEntry("/META-INF/resources/WEB-INF/tags");
+            if (metaInfResourcesEntry != null) {
+                Container tagsContainer = metaInfResourcesEntry.adapt(Container.class);
+                if (tagsContainer != null) {
+                    // Use /WEB-INF/tags as the cache key so JSPs can find them with tagdir="/WEB-INF/tags"
+                    // The actual path in the JAR is /META-INF/resources/WEB-INF/tags
+                    String tagsDirInJar = "/META-INF/resources/WEB-INF/tags";
+                    String tagsDirKey = "/WEB-INF/tags";  // Cache key that JSPs will use
+                    if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.FINE)) {
+                        logger.logp(Level.FINE, CLASS_NAME, "loadTagFilesFromJar", "Loading tag files from JAR: " + jarName + ", path: " + tagsDirInJar);
+                    }
+                    loadTagFilesFromJarRecursive(tagsContainer, tagsDirInJar, tagsDirKey, jarName);
+                }
+            }
+        } catch (UnableToAdaptException e) {
+            if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.WARNING)) {
+                logger.logp(Level.WARNING, CLASS_NAME, "loadTagFilesFromJar", "Failed to load tag files from JAR: " + jarName, e);
+            }
+        }
+    }
+    
+    /**
+     * Recursively loads tag files from a container representing a tags directory in a JAR.
+     */
+    @SuppressWarnings("unchecked")
+    private void loadTagFilesFromJarRecursive(Container tagsContainer, String tagsDirInJar, String tagsDirKey, String jarName) throws JspCoreException {
+        Set<String> tagFileSet = new HashSet<String>();
+        getFilesFromContainer(tagFileSet, tagsContainer);
+        
+        ArrayList<TagFileInfo> list = new ArrayList<TagFileInfo>();
+        TagLibraryInfoImpl tli = null;
+        boolean succeeded = true;
+        
+        if (tagFileSet != null && !tagFileSet.isEmpty()) {
+            ImplicitTldParser implicitTldParser = new ImplicitTldParser(ctxt, configManager, false);
+            
+            // First, check for implicit.tld
+            for (String resourcePath : tagFileSet) {
+                if (resourcePath.endsWith("/implicit.tld")) {
+                    try {
+                        JspInputSource inputSource = ctxt.getJspInputSourceFactory().createJspInputSource(resourcePath);
+                        tli = implicitTldParser.parseTLD(inputSource, "jar");
+                        if (tli != null) {
+                            if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.FINE)) {
+                                logger.logp(Level.FINE, CLASS_NAME, "loadTagFilesFromJarRecursive", "Got TagLibraryInfoImpl for [{0}] from JAR [{1}], TLD file [{2}]", new Object[]{tagsDirKey, jarName, resourcePath});
+                            }
+                        }
+                    } catch (JspCoreException e) {
+                        if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.WARNING)) {
+                            logger.logp(Level.WARNING, CLASS_NAME, "loadTagFilesFromJarRecursive", "JAR tagfile directory tld failed to load for resourcePath=[" + resourcePath + "] in JAR [" + jarName + "]", e);
+                            logger.logp(Level.WARNING, CLASS_NAME, "loadTagFilesFromJarRecursive", "tagfiles in " + tagsDirKey + " from JAR " + jarName + " are not loaded");
+                        }
+                        succeeded = false;
+                    }
+                }
+            }
+            
+            if (succeeded) {
+                // Create input source from the JAR container, not from the context
+                // Use the actual JAR path for the input source
+                JspInputSource inputSource = getInputSource(tagsContainer, tagsDirInJar, null, null);
+                // But use the /WEB-INF/tags key for the ImplicitTagLibraryInfoImpl so JSPs can find it
+                TagLibraryInfoImpl implicitTli = new ImplicitTagLibraryInfoImpl(tagsDirKey, inputSource);
+                
+                if (tli != null) {
+                    if (tli.getRequiredVersion() != null)
+                        implicitTli.setRequiredVersion(tli.getRequiredVersion());
+                    if (tli.getTlibversion() != null)
+                        implicitTli.setTlibversion(tli.getTlibversion());
+                }
+                
+                // Process .tag and .tagx files
+                for (String resourcePath : tagFileSet) {
+                    if (resourcePath.endsWith(".tag") || resourcePath.endsWith(".tagx")) {
+                        try {
+                            if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.FINEST)) {
+                                logger.logp(Level.FINEST, CLASS_NAME, "loadTagFilesFromJarRecursive", "about to do tagfilescan for = [" + resourcePath + "] from JAR [" + jarName + "]");
+                            }
+                            
+                            // Create input source from the JAR container for the specific tag file
+                            JspInputSource tagFileInputSource = getInputSource(tagsContainer, resourcePath, null, null);
+                            JspTranslator jspTranslator = JspTranslatorFactory.getFactory().createTranslator(TAGFILE_SCAN_ID,
+                                                                                                            tagFileInputSource,
+                                                                                                            ctxt,
+                                                                                                            configManager.createJspConfiguration(),
+                                                                                                            jspOptions,
+                                                                                                            implicitTagLibPrefixMap);
+                            
+                            JspVisitorInputMap inputMap = new JspVisitorInputMap();
+                            inputMap.put("TagLibraryInfo", implicitTli);
+                            String name = resourcePath.substring(resourcePath.lastIndexOf('/') + 1);
+                            // Remove .tag or .tagx extension
+                            if (name.endsWith(".tag")) {
+                                name = name.substring(0, name.length() - 4);
+                            } else if (name.endsWith(".tagx")) {
+                                name = name.substring(0, name.length() - 5);
+                            }
+                            inputMap.put("TagFileName", name);
+                            inputMap.put("TagFilePath", resourcePath);
+                            
+                            HashMap results = jspTranslator.processVisitors(inputMap);
+                            TagFileScanResult result = (TagFileScanResult) results.get("TagFileScan");
+                            TagFileInfo tfi = new TagFileInfo(name, resourcePath, result.getTagInfo());
+                            list.add(tfi);
+                            
+                            if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.FINEST)) {
+                                logger.logp(Level.FINEST, CLASS_NAME, "loadTagFilesFromJarRecursive", "TagLibraryCache TagFileInfo tfi=[" + tfi + "] from JAR [" + jarName + "]");
+                            }
+                        } catch (JspCoreException e) {
+                            if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.WARNING)) {
+                                logger.logp(Level.WARNING, CLASS_NAME, "loadTagFilesFromJarRecursive", "JAR tagfile failed to scan=[" + resourcePath + "] in JAR [" + jarName + "]", e);
+                            }
+                        }
+                    }
+                }
+                
+                if (list.size() > 0) {
+                    implicitTli.setTagFiles(list);
+                    list.clear();
+                    implicitTli.setFunctions(Collections.emptyList());
+                    implicitTli.setTags(Collections.emptyList());
+                    
+                    if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.FINE)) {
+                        logger.logp(Level.FINE, CLASS_NAME, "loadTagFilesFromJarRecursive", "Adding ImplicitTagLibraryInfoImpl for [{0}] from JAR [{1}]", new Object[]{tagsDirKey, jarName});
+                        logger.logp(Level.FINE, CLASS_NAME, "loadTagFilesFromJarRecursive", "  ImplicitTagLibraryInfoImpl=[{0}]", implicitTli);
+                    }
+                    put(tagsDirKey, implicitTli);
+                } else if (tli != null) {
+                    put(tagsDirKey, tli);
+                }
+            }
+        }
+        
+        // Recursively process subdirectories
+        try {
+            for (com.ibm.wsspi.adaptable.module.Entry entry : tagsContainer) {
+                Container subContainer = entry.adapt(Container.class);
+                if (subContainer != null && entry.getSize() == 0) {
+                    String subDir = entry.getName();
+                    if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.FINE)) {
+                        logger.logp(Level.FINE, CLASS_NAME, "loadTagFilesFromJarRecursive", "Recursing into subdirectory: " + subDir + " in JAR: " + jarName);
+                    }
+                    // For subdirectories, use the subdirectory name as both the JAR path and cache key
+                    loadTagFilesFromJarRecursive(subContainer, subDir, subDir, jarName);
+                }
+            }
+        } catch (UnableToAdaptException e) {
+            if (com.ibm.ejs.ras.TraceComponent.isAnyTracingEnabled() && logger.isLoggable(Level.WARNING)) {
+                logger.logp(Level.WARNING, CLASS_NAME, "loadTagFilesFromJarRecursive", "Failed to process subdirectories in " + tagsDirKey + " from JAR " + jarName, e);
             }
         }
     }

--- a/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/resource/TagFileResourcesImpl.java
+++ b/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/resource/TagFileResourcesImpl.java
@@ -64,10 +64,19 @@ public class TagFileResourcesImpl extends ResourcesImpl implements TagFileResour
         if (tfi.getPath().startsWith("/WEB-INF/tags")) {
             tagFilePath = tfi.getPath().substring(tfi.getPath().indexOf("/WEB-INF/tags") + 13);
         }
+        else if (tfi.getPath().startsWith("/META-INF/resources/WEB-INF/tags")) {
+            // Handle tag files from JARs: /META-INF/resources/WEB-INF/tags/...
+            tagFilePath = tfi.getPath().substring(tfi.getPath().indexOf("/META-INF/resources/WEB-INF/tags") + 32);
+        }
         else if (tfi.getPath().startsWith("/META-INF/tags")) {
             tagFilePath = tfi.getPath().substring(tfi.getPath().indexOf("/META-INF/tags") + 14);
         }
-        tagFilePath = tagFilePath.substring(0, tagFilePath.lastIndexOf("/"));
+        
+        if (tagFilePath != null && tagFilePath.contains("/")) {
+            tagFilePath = tagFilePath.substring(0, tagFilePath.lastIndexOf("/"));
+        } else if (tagFilePath != null) {
+            tagFilePath = "";  // No subdirectory, just the tags directory itself
+        }
         
         //PM70267 start
         if (tagFilePath.indexOf("-") > -1) {

--- a/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/GenerateTagFileVisitor.java
+++ b/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/GenerateTagFileVisitor.java
@@ -181,10 +181,18 @@ public class GenerateTagFileVisitor extends GenerateVisitor {
         String tagFilePath = null;
         if (tfi.getPath().startsWith("/WEB-INF/tags")) {
             tagFilePath = tfi.getPath().substring(tfi.getPath().indexOf("/WEB-INF/tags") + 13);
+        } else if (tfi.getPath().startsWith("/META-INF/resources/WEB-INF/tags")) {
+            // Handle tag files from JARs: /META-INF/resources/WEB-INF/tags/...
+            tagFilePath = tfi.getPath().substring(tfi.getPath().indexOf("/META-INF/resources/WEB-INF/tags") + 32);
         } else if (tfi.getPath().startsWith("/META-INF/tags")) {
             tagFilePath = tfi.getPath().substring(tfi.getPath().indexOf("/META-INF/tags") + 14);
         }
-        tagFilePath = tagFilePath.substring(0, tagFilePath.lastIndexOf("/"));
+        
+        if (tagFilePath != null && tagFilePath.contains("/")) {
+            tagFilePath = tagFilePath.substring(0, tagFilePath.lastIndexOf("/"));
+        } else if (tagFilePath != null) {
+            tagFilePath = "";  // No subdirectory, just the tags directory itself
+        }
         //PM70267 start
         if (tagFilePath.indexOf("-") > -1) {
             tagFilePath = com.ibm.ws.jsp.translator.utils.NameMangler.handlePackageName(tagFilePath);

--- a/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/tagfilescan/TagFileScanVisitor.java
+++ b/dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/tagfilescan/TagFileScanVisitor.java
@@ -91,6 +91,9 @@ public class TagFileScanVisitor extends JspVisitor {
         String dirName = tagFilePath.substring(0, tagFilePath.lastIndexOf('/'));
         if (dirName.startsWith("/WEB-INF/tags")) {
             dirName = dirName.substring(dirName.indexOf("/WEB-INF/tags") + 13);
+        } else if (dirName.startsWith("/META-INF/resources/WEB-INF/tags")) {
+            // Handle tag files from JARs: /META-INF/resources/WEB-INF/tags/...
+            dirName = dirName.substring(dirName.indexOf("/META-INF/resources/WEB-INF/tags") + 32);
         } else if (dirName.startsWith("/META-INF/tags")) {
             dirName = dirName.substring(dirName.indexOf("/META-INF/tags") + 14);
         }

--- a/dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/JspEngineElement.java
+++ b/dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/JspEngineElement.java
@@ -25,6 +25,7 @@ public class JspEngineElement extends ConfigElement {
     private Boolean keepGenerated;
     private Boolean useJDKCompiler;
     private String prepareJSPs;
+    private Boolean loadTagFilesFromJars;
 
     /**
      * @return the useStringCast
@@ -144,6 +145,18 @@ public class JspEngineElement extends ConfigElement {
         this.prepareJSPs = prepareJSPs;
     }
 
+    /**
+     * @return the loadTagFilesFromJars
+     */
+    public Boolean isLoadTagFilesFromJars() {
+        return loadTagFilesFromJars;
+    }
+
+    @XmlAttribute(name = "loadTagFilesFromJars")
+    public void setLoadTagFilesFromJars(Boolean b) {
+        this.loadTagFilesFromJars = b;
+    }
+
     @Override
     public String toString() {
         StringBuffer buf = new StringBuffer("JspElement{");
@@ -165,8 +178,10 @@ public class JspEngineElement extends ConfigElement {
             buf.append("keepGenerated=\"" + keepGenerated + "\" ");
         if (useJDKCompiler != null)
             buf.append("useJDKCompiler=\"" + useJDKCompiler + "\" ");
-        if (useJDKCompiler != null)
+        if (prepareJSPs != null)
             buf.append("prepareJSPs=\"" + prepareJSPs + "\" ");
+        if (loadTagFilesFromJars != null)
+            buf.append("loadTagFilesFromJars=\"" + loadTagFilesFromJars + "\" ");
 
         buf.append("}");
         return buf.toString();

